third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,599 @@
+    pub fn iter_dfs_left(&self, visit_all: bool) -> impl Iterator<Item = BlockId> + '_ {
+        DFSLeft::new(&self.successors, self.entry_block, visit_all)
+    }
+
+    /// Generates code from the control flow graph
+    /// `visit_all`: determines whether to generate code for unreachable blocks
+    pub fn gen_code(self, visit_all: bool) -> Vec<Bytecode> {
+        let mut generated = Vec::new();
+        let mut iter_dfs_left = self.iter_dfs_left(visit_all).peekable();
+        while let Some(block) = iter_dfs_left.next() {
+            if self.is_trivial_block(block) {
+                continue;
+            }
+            // TODO:
+            // can't use `.remove` instead to avoid copying,
+            // because the following may look at visited block
+            let mut code_block = self.code_blocks.get(&block).expect("code block").clone();
+            code_block = self.gen_code_for_block(code_block, block, iter_dfs_left.peek());
+            generated.append(&mut code_block);
+        }
+        generated
+    }
+
+    /// Generates code for block. The way we generate code effectively does the transformation 3 in the module doc.
+    fn gen_code_for_block(
+        &self,
+        mut code_block: Vec<Bytecode>,
+        block: BlockId,
+        next_block_to_visit: Option<&BlockId>,
+    ) -> Vec<Bytecode> {
+        // since we may generate the blocks in a different order, we may need to add explicit jump or eliminate unnecessary jump.
+        if Self::falls_to_next_block(&code_block) {
+            // if we have block 0 followed by block 1 without jump/branch
+            // and we don't visit block 1 after block 0, then we have to add an explicit jump
+            let succ_block = self.get_the_non_trivial_successor(block);
+            if next_block_to_visit.is_none() || *next_block_to_visit.unwrap() != succ_block {
+                self.add_explicit_jump(&mut code_block, succ_block);
+            }
+        } else if matches!(
+            code_block.last().expect("last instruction"),
+            Bytecode::Jump(..)
+        ) {
+            // no need to jump to the next instruction
+            let succ_block = self.get_the_non_trivial_successor(block);
+            if let Some(next_to_vist) = next_block_to_visit {
+                if *next_to_vist == succ_block {
+                    remove_tail_jump(&mut code_block);
+                }
+            }
+        }
+        code_block
+    }
+
+    /// Checks whether a block falls to the next block without jump, branch, abort, or return;
+    /// i.e., the block is followed by the next in the original code
+    fn falls_to_next_block(code: &[Bytecode]) -> bool {
+        let last_instr = code.last().expect("last instr");
+        !last_instr.is_always_branching()
+    }
+
+    /// Gets the only successor of `block`; panics if there is no successors.
+    /// (May not panic in the release version due to the use of debug_assert)
```

Comment:
`*successors.first().expect("successor block")` will panic, even in release mode, if there *are* no successors.

Reply:
done